client/faucet/src/main.rs

```diff
@@ -0,0 +1,339 @@
+}
+
+struct OptFmt<T>(Option<T>);
+
+impl<T: fmt::Display> fmt::Display for OptFmt<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if let Some(ref t) = self.0 {
+            fmt::Display::fmt(t, f)
+        } else {
+            f.write_str("-")
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ServerInternalError(String);
+impl warp::reject::Reject for ServerInternalError {}
+
+#[cfg(test)]
+mod tests {
+    use crate::routes;
+    use libra_faucet::mint;
+    use std::{collections::HashMap, sync::Arc};
+    use warp::Filter;
+
+    fn setup() -> (tokio::runtime::Runtime, Arc<mint::Service>) {
+        setup_with_accounts(genesis_accounts())
+    }
+
+    fn setup_with_accounts(
+        accounts: HashMap<String, serde_json::Value>,
+    ) -> (tokio::runtime::Runtime, Arc<mint::Service>) {
+        let f = tempfile::NamedTempFile::new()
+            .unwrap()
+            .into_temp_path()
+            .to_path_buf();
+        generate_key::generate_and_save_key(&f);
+
+        let chain_id = libra_types::chain_id::ChainId::test();
+        let rt = tokio::runtime::Runtime::new().unwrap();
+
+        let accounts_arc = Arc::new(accounts);
+        let stub = warp::any()
+            .and(warp::body::json())
+            .map(move |req: serde_json::Value| {
+                let resp = handle_request(req, chain_id, Arc::clone(&accounts_arc));
+                Ok(warp::reply::json(&resp))
+            });
+        let port = libra_config::utils::get_available_port();
+        let server = rt.enter(move || warp::serve(stub).bind(([127, 0, 0, 1], port)));
+        rt.handle().spawn(server);
+
+        let service = mint::Service::new(
+            format!("http://localhost:{}/v1", port),
+            chain_id,
+            f.to_str().unwrap().to_owned(),
+        );
+        (rt, Arc::new(service))
+    }
+
+    #[test]
+    fn test_healthy() {
+        let (mut rt, service) = setup();
+        let filter = routes(service);
+        let resp = rt.block_on(async {
+            warp::test::request()
+                .method("GET")
+                .path("/-/healthy")
+                .reply(&filter)
+                .await
+        });
+        assert_eq!(resp.status(), 200);
+        assert_eq!(resp.body(), "libra-faucet:ok");
+    }
+
+    #[test]
+    fn test_mint() {
+        let (mut rt, service) = setup();
+        let filter = routes(service);
+
+        let auth_key = "459c77a38803bd53f3adee52703810e3a74fd7c46952c497e75afb0a7932586d";
+        let resp = rt.block_on(async {
+            warp::test::request()
+                .method("POST")
+                .path(
+                    format!(
+                        "/mint?auth_key={}&amount=1000000&currency_code=LBR",
+                        auth_key
+                    )
+                    .as_str(),
+                )
+                .reply(&filter)
+                .await
+        });
+        assert_eq!(resp.body(), "2389"); // 2388+1
+    }
+
+    #[test]
```

Comment:
You can use the `#[tokio::test]` attribute to make test functions "async" so that you don't have to create a runtime for each test.

Reply:
thanks for the hint, look better now